### DISCUSSION TITLE
AI: Generate Description with Clip Interrogator

### DISCUSSION THREAD

**Discussion by sfxworks** [photoprism/photoprism#3419] [Ideas]
![image](https://github.com/photoprism/photoprism/assets/5921035/4ca182e5-fe3f-476f-9162-931e190b1e5d)
I was looking around a lot for a way to handle and autotag images using something better than the tensor model. I made a stable diffusion interrogator wrapper/endpoint at https://github.com/sfxworks/interrogator-http and a script that (roughly) calls the API to add a description and label the images it goes through. You'll need your session ID, as well as some additional token I found but couldn't find in documentation as noted by my `t=small-token-here` example.

You can use multiple containers. Though it could use some fine tuning when you do so since the iteration doesn't handle errors yet.

I am currently using model ViT-H-14/laion2b_s32b_b79k that runs against Stable Diffusion v2. More information here https://github.com/pharmapsychotic/clip-interrogator

This is a rough script. A lot more work needs to be done such as handle labels a bit better (maybe join them instead of just by word), get video image thumbnails instead of skipping them, and other methods. Note that it does take about 2m to interrogate an image to get these sort of results, and I currently have it running on a GTX 1070 with some performance args to reduce memory (which also reduce quality). So your mileage may vary. 

```python
import argparse
import requests
from PIL import Image
from io import BytesIO
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
import base64
from concurrent.futures import ThreadPoolExecutor
import itertools

# Set up argument parser
parser = argparse.ArgumentParser(description='Process some integers.')
parser.add_argument('offset', type=int, help='The starting index of the photos to process')
parser.add_argument('count', type=int, help='The ending index of the photos to process')

# List of interrogator IPs
server_ips = ['192.168.6.16'] #, '10.0.8.88', '10.0.8.120', '10.0.8.115', '10.0.8.46']


args = parser.parse_args()

# Download the set of English stop words
nltk.download('punkt')
nltk.download('stopwords')
stop_words = set(stopwords.words('english'))

# Set up PhotoPrism API details
photoprism_url = "https://photoprism.url.here/api/v1"
headers = {"X-Session-ID": "seesion-id-here"}

# Fetch the list of photo UUIDs from PhotoPrism
response = requests.get(f"{photoprism_url}/photos?count={args.count}&offset={args.offset}", headers=headers)
photo_list = response.json()

# Define function to process a single photo
def process_photo(photo, ip):
    file_name = photo.get('FileName')
    if file_name and file_name.endswith('.mp4'):
        print("Skipping MP4 file: " + file_name)
        return
    
    print("Processing photo against IP: " + ip)
    print("Photo: " + photo['Hash'])
    print("Photo: " + photo['UID'])

    # Use the provided IP for this request
    url = f"http://{ip}/interrogate"

    # Fetch the image data
    image_url = f"{photoprism_url}/dl/{photo['Hash']}?t=small-token-here"
    print("Using: " + image_url)
    response = requests.get(image_url, headers=headers)
    image_data = base64.b64encode(response.content).decode('utf-8')
    print("Got image data response.")
    
    # Send the image data to the interrogator endpoint
    print("Sending to interrogator")
    response = requests.post(url, json={'image': image_data}, timeout=600)  # Wait up to 600 seconds (10 minutes)
    label_sentence = response.json()['label']

    print("Labels before:")
    print(label_sentence)

    description_url = f"{photoprism_url}/photos/{photo['UID']}"

    payload = {
      "Description": label_sentence,
      "DescriptionSrc": "manual",
      "Notes" :"Description and tags added by ViT-H-14/laion2b_s32b_b79k",
      "NotesSrc": "manual",
    }

    response = requests.put(description_url, headers=headers, json=payload)
    if response.status_code != 200:
        print(f"Failed to add description '{label_sentence}' to photo {photo['UID']}: {response.content}")

    # Tokenize the label and remove stop words
    label_words = word_tokenize(label_sentence)
    label_keywords = [word for word in label_words if word not in stop_words and word != ',' and word != '!']
    print("Labels after:")
    print(label_keywords)

    # Add each keyword as a label to the photo in PhotoPrism
    for keyword in label_keywords:
        # Prepare the JSON payload for the PhotoPrism API
        payload = {
          "Name": keyword,
          "Priority": 10,  # Adjust this as needed
        }

        # Add the label to the photo
        label_url = f"{photoprism_url}/photos/{photo['UID']}/label"
        response = requests.post(label_url, headers=headers, json=payload)
        
        if response.status_code != 200:
            print(f"Failed to add label '{keyword}' to photo {photo['UID']}: {response.content}")

# Create a cyclic iterator for the server IPs
cyclic_ips = itertools.cycle(server_ips)

# In the ThreadPoolExecutor block:
with ThreadPoolExecutor(max_workers=1) as executor:
    # Map the process_photo function to the photo_list and cyclic_ips
    executor.map(process_photo, photo_list, cyclic_ips)
```

**Comment by lastzero** [comment]
The generated description looks fantastic. Our scarcest resource is time, as we already have many other feature requests on our list. So if we can limit that by having tested API specs and dependencies (like a ready-to-use Docker image), chances are we can integrate this to make it available to all users. Also, it would be good to know the hardware requirements (CPU cores and architecture, AVX, AVX2, AVX512, GPU, RAM, disk size,...). Thank you! ❤️

**Reply by lastzero** [reply]
I've created the follwing issue from this:
- #3438

**Comment by sfxworks** [comment]
Heya! Sorry for the delay! 
Hardware wise, 
Running a GTX 1070Ti (the worst of the worst) that has 8G of VRAM
CPU and Memory wise it uses very little since it's just decoding the b64 image and throwing it to the gpu. 
Using the nvidia container toolkit here. I haven't tested this on any AMD gpus but I can try to do soon as I would like to use more of those instead.

**Reply by lastzero** [reply]
Is there a realistic chance of this running on ARM64 or an Intel/AMD CPU without a dedicated GPU?

**Reply by sfxworks** [reply]
Maybe with some work against some of the libraries used in stablediffusion web ui repo, but not something that would be easy. The RAM requirement goes way up too. Vram from 8G to 40G to do text to image for example.

**Reply by lastzero** [reply]
I see, so it should be optional and requires documentation to set up properly 👍

**Reply by lastzero** [reply]
Note that we plan to add configurable hooks and/or support for Lua scripts. This would make it easier to use additional machine learning models and allow us to merge related PRs faster, especially if the additional models run in separate containers so that our base image doesn't need to be updated. An intern joining us next month may be able to help with this.
